include/cocaine/signal.hpp

```diff
@@ -0,0 +1,153 @@
+/*
+    Copyright (c) 2011-2015 Anton Matveenko <antmat@yandex-team.ru>
+    Copyright (c) 2011-2015 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef COCAINE_SIGNAL_SET_HPP
+#define COCAINE_SIGNAL_SET_HPP
+
+#include <cocaine/locked_ptr.hpp>
+
+#include <atomic>
+#include <vector>
+#include <limits.h>
+#include <map>
+#include <functional>
+#include <system_error>
+#include <signal.h>
+#include <array>
+
+
+namespace cocaine { namespace signal {
+
+class cancellation_t {
+public:
+    // Call cancellation callback if present, was not called before and was not detached.
+    // Return true if callback was called
+    bool
+    cancel();
+
+    // Detach cancellation token so the destructor or cancel() would not call cancellation callback.
+    // Usually used right after async wait, f.e. signal_handler.async_wait(cb).detach();
+    // Return if callback was sucessfully detached(was not cancelled before, detached before and callback is present)
+    bool
+    detach();
+
+    cancellation_t();
+    cancellation_t(std::function<void()> cancellation_callback);
+    cancellation_t(const cancellation_t&) = delete;
+    cancellation_t& operator=(const cancellation_t&) = delete;
+    cancellation_t(cancellation_t&&);
+    cancellation_t& operator=(cancellation_t&&);
+    ~cancellation_t();
+
+private:
+    std::atomic_flag cancelled;
+    std::function<void()> cancellation_callback;
+};
+
+class handler_i {
```

Comment:
`*_i`? That's something new, don't do that. Uncle Bob says that the `_i` as a suffix of interfaces is another backwards convention. When you are passed a reference/pointer to an object, you should expect it to be an interface. Either you don't need an interface or rename the implementation.